### PR TITLE
Release v1.0.0 beta02

### PR DESCRIPTION
**Pull request by bryanapellanes-okta**
Contains the following:
- OIDC Logout implementation
- Xamarin.Forms compatibility
- Nuget packages:
  - Okta.Xamarin.1.0.0-beta02.nupkg
  - Okta.Xamarin.Android.1.0.0-beta02.nupkg
  - Okta.Xamarin.iOS.1.0.0-beta02.nupkg

Additional work for future PR(s):
- Move UITests into `samples-xamarin` repo
- Fix formatting from legacy sample code
- Add code documentation and file headers
- Update readme to include information on added convenience components.

## Issue \#
[OKTA-342886](https://oktainc.atlassian.net/browse/OKTA-342886)

## Code

- [x] Unit test(s)
- [x] Implementation

### CODE REVIEW

File: Okta.Xamarin/Okta.Xamarin.Test/OidcClientShould.cs [hunk 1]
```diff
@@ -5,229 +5,267 @@
...
-			});
-
-			await Assert.ThrowsAsync<OAuthException>(() => client.SignInWithBrowserAsync());
+			Assert.True(string.IsNullOrEmpty(client.State_Internal));
+			Assert.True(string.IsNullOrEmpty(client.CodeChallenge_Internal));
+			Assert.True(string.IsNullOrEmpty(client.CodeVerifier_Internal));
+			
+			client.SignOutOfOktaAsync(new OktaState("testAccessToken", "testTokenType"));
+			
+			Assert.False(string.IsNullOrEmpty(client.State_Internal));
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.Test/OidcClientShould.cs:246]: I'm not quite following this test. What exactly are we testing here?

File: Okta.Xamarin/Okta.Xamarin.Test/OidcClientShould.cs [hunk 2]
```diff
@@ -5,229 +5,267 @@
...
+        [Fact]
+        public void FailWithInvalidConfig()
+        {
+            Assert.Throws<ArgumentNullException>(() => new OidcClient(new OktaConfig()));
+        }
+
+        [Fact]
+        public void GetCorrectAuthUrl()
+        {
+            OidcClient client = new OidcClient(new OktaConfig("testoktaid", "https://dev-00000.oktapreview.com", "com.test:/redirect+&TEST!@url%20Encode#*^(0)", "com.test:/logout") { Scope = "test hello test_scope" });
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.Test/OidcClientShould.cs:27]: Can we also add a test to verify that the sign-out URL is well-formed ? It should be pretty similar to this test.

File: Okta.Xamarin/Okta.Xamarin.Test/OidcClientShould.cs (hunk 1 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.Test/OidcClientShould.cs:246]: This is making sure the internal state of the client gets set by the call to signout.  The State, CodeChallenge and CodeVerifier are private to the client but exposed as "_Internal" for testing.  Those values are needed for the background redirect flows.

File: Okta.Xamarin/Okta.Xamarin.Test/OidcClientShould.cs (hunk 2 above)
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.Test/OidcClientShould.cs:27]: I'll add that to my list of todos.

File: Okta.Xamarin/Okta.Xamarin.Test/OktaContextShould.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		public void RaiseSignInEventsOnSignIn()
+		{
+			OidcClient client = new OidcClient(new OktaConfig("testoktaid", "https://dev-00000.oktapreview.com", "com.test:/redirect", "com.test:/logout"));
+			OktaContext.Init(client);
+			bool? signInStartedEventRaised = false;
+			bool? signInCompletedEventRaised = false;
+			OktaContext.AddSignInStartedListener((sender, eventArgs) => signInStartedEventRaised = true);
+			OktaContext.AddSignInCompletedListener((sender, eventArgs) => signInCompletedEventRaised = true);
+			
+			OktaContext.Current.SignIn();
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.Test/OktaContextShould.cs:20]: Probably to handle in a separate PR, but it seems we use both `Sign In` and `Login` for the same action. I'd be great we use one of those for consistency.

Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.Test/OktaContextShould.cs:20]: Agreed.

File: Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        /// <returns>The computed Access Token Url used for retrieving a token</returns>
+        public string GetAccessTokenUrl()
+        {
+            return $"{OktaDomain}/oauth2/{AuthorizationServerId}/v1/token";
+        }
+
+		public string GetLogoutUrl()
+		{
+			return $"{OktaDomain}/oauth2/{AuthorizationServerId}/v1/logout";
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs:55]: Have you tested logout using the Okta org (AuthorizationServerId is "" in that case)? IIRC the URL is different. Take a look at this: https://github.com/okta/okta-aspnet/blob/5a2f459934e004a8e1211928eb548e1c1be3f378/Okta.AspNet.Abstractions/UrlHelper.cs#L14

Comment (laura-rodriguez) [reply Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs:55]: It actually seems to be a legacy issue, all other URLs (token, authorize) don't take the Okta Org Authz server into account. You might want to revisit this in the future.

Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.UITest.Shared/TestOktaConfig.cs:55]: I'll plan to add a test for that.

File: Okta.Xamarin/Okta.Xamarin.iOS/OktaViewController.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+		public OktaConfig OktaConfig { get; set; }
+
+		public override void ViewDidLoad()
+		{
+			base.ViewDidLoad();
+			// Perform any additional setup after loading the view, typically from a nib.
+
+			OktaConfig config = OktaConfig.LoadFromPList("OktaConfig.plist");
+			// TODO: log telemetry data
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin.iOS/OktaViewController.cs:54]: Are we tracking user-agent info?

Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin.iOS/OktaViewController.cs:54]: When we're ready we will likely integrate the other Sdk's to make those calls when we decide that's appropriate.

File: Okta.Xamarin/Okta.Xamarin/OidcClient.cs [hunk 6]
```diff
@@ -13,279 +13,362 @@
...
+        public Task<OktaState> SignOutOfOktaAsync(OktaState stateManager)
+        {
+            validator.Validate(this.Config);
+            this.currentTask = new TaskCompletionSource<OktaState>();
+            if (!stateManager.IsAuthenticated)
+            {
+				this.currentTask.SetResult(stateManager);
+				return this.currentTask.Task;
+			}
+			this.GenerateStateCodeVerifierAndChallenge();
```
Comment (laura-rodriguez) [inline Okta.Xamarin/Okta.Xamarin/OidcClient.cs:73]: Question for my own education. Do we also need to challenge when logging out when using PKCE?

Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/OidcClient.cs:73]: I don't think it is strictly required, just a good idea; however I'm not certain either way.

File: Okta.Xamarin/Okta.Xamarin/OidcClient.cs [hunk 7]
```diff
@@ -13,279 +13,362 @@
...
+                byte[] data = new byte[byteCount];
+                rng.GetBytes(data);
+
+                return Microsoft.IdentityModel.Tokens.Base64UrlEncoder.Encode(data);
+            }
+        }
+
+        private string GenerateLogoutUrl(LogoutOptions logoutOptions)
+        {
+			var baseUri = new Uri(this.Config.GetLogoutUrl());
```
[the remark this answers is not in the document]
Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/OidcClient.cs:247]: On the next line, the call to logoutOptions.ToQueryString will append the relevant values to the querystring.

**Review by laura-rodriguez — APPROVED** [review]
Approving this so the remaining work is easy to review. Please, before publishing a new version let's make sure we are following the PKCE spec since we are implementing it manually. Also, please consider adding a changelog to keep track of all the changes you've been working on.

File: Okta.Xamarin/Okta.Xamarin/Base64UrlEncoder.cs [hunk 8]
```diff
@@ -30,109 +30,109 @@
...
-					str += base64PadCharacter;
-					break;
-			}
+            // check for padding
+            switch (str.Length % 4)
+            {
+                case 0:
+                    // No pad chars in this case
+                    break;
+                case 2:
```
Comment (andriizhegurov-okta) [inline Okta.Xamarin/Okta.Xamarin/Base64UrlEncoder.cs:115]: Maybe `case 1` is missed here.

Comment (bryanapellanes-okta) [reply Okta.Xamarin/Okta.Xamarin/Base64UrlEncoder.cs:115]: This file is legacy and should probably be replaced, at some point, with an established bas364 encoder.

**Pull request merged**